app/services/db/identity/sp_active_user_counts_within_iaa_window.rb

```diff
@@ -0,0 +1,60 @@
+module Db
+  module Identity
+    # Similar to SpActiveUserCounts, but it limits dates to within active IAA windows
+    class SpActiveUserCountsWithinIaaWindow
+      def self.call
+        service_providers = ServiceProvider.arel_table
+        identities = ServiceProviderIdentity.arel_table
+
+        ial1_within_iaa = between(
+          identities[:last_ial1_authenticated_at],
+          service_providers[:iaa_start_date],
+          service_providers[:iaa_end_date],
+        )
+
+        ial2_within_iaa = between(
+          identities[:last_ial2_authenticated_at],
+          service_providers[:iaa_start_date],
+          service_providers[:iaa_end_date],
+        )
+
+        sql = ServiceProvider.
+          select(
+            service_providers[:issuer],
+            service_providers[:app_id].maximum.as('app_id'),
+            service_providers[:iaa].maximum.as('iaa'),
+            service_providers[:iaa_start_date].minimum.as('iaa_start_date'),
+            service_providers[:iaa_end_date].maximum.as('iaa_end_date'),
+            sum_if(ial1_within_iaa).as('total_ial1_active'),
+            sum_if(ial2_within_iaa).as('total_ial2_active'),
+          ).joins(:identities).
+          where(
+            ial1_within_iaa.or(ial2_within_iaa),
+          ).group(service_providers[:issuer]).
+          order(service_providers[:issuer]).
+          to_sql
+
+        ActiveRecord::Base.connection.execute(sql)
+      end
+
+      # Builds "value BETWEEN range_start AND range_end"
+      # We need to do this because Arel's value#between() only works on values
+      # that can go in the stdlib Range class
+      def self.between(value, range_start, range_end)
```

Comment:
This and the `sum_if` are harder to follow for me, though I don't know how it would compare to the SQL

Reply:
Unfortunately this will not work using the identities table.  The reason is the last_authenticated_at does not contain a history.  If I am active over the POP and then I'm active after the POP it will not show me active in the POP.

Reply:
@stevegsa do we have this data in another table?

Reply:
If you need to use end date the best we have is monthly_sp_auth_counts which can give you an arbitrary range but it does not have day resolution.  Since we bill monthly it's generally not an issue

Reply:
> This and the `sum_if` are harder to follow for me, though I don't know how it would compare to the SQL

Here's the resulting SQL (with some newlines added)

```sql
SELECT
  "service_providers"."issuer",
  MAX("service_providers"."app_id") AS app_id,
  MAX("service_providers"."iaa") AS iaa,
  MIN("service_providers"."iaa_start_date") AS iaa_start_date,
  MAX("service_providers"."iaa_end_date") AS iaa_end_date,
  SUM(
    CASE
    WHEN "identities"."last_ial1_authenticated_at" BETWEEN "service_providers"."iaa_start_date" AND "service_providers"."iaa_end_date"
    THEN 1
    ELSE 0
    END
  ) AS total_ial1_active,
  SUM(
    CASE
    WHEN "identities"."last_ial2_authenticated_at" BETWEEN "service_providers"."iaa_start_date" AND "service_providers"."iaa_end_date"
    THEN 1
    ELSE 0
    END
  ) AS total_ial2_active
FROM "service_providers"
INNER JOIN "identities" ON "identities"."service_provider" = "service_providers"."issuer"
WHERE (
  "identities"."last_ial1_authenticated_at" BETWEEN "service_providers"."iaa_start_date" AND "service_providers"."iaa_end_date"
  OR "identities"."last_ial2_authenticated_at" BETWEEN "service_providers"."iaa_start_date" AND "service_providers"."iaa_end_date"
)
GROUP BY "service_providers"."issuer" ORDER BY "service_providers"."issuer"
```

Reply:
SQL looks good to me, makes it easier to run in `psql` too if we have the query handy 🙂

Reply:
all you gotta do its `puts sql` tho from the code 😬 (which is what I did to generate this)

Reply:
> If you need to use end date the best we have is monthly_sp_auth_counts which can give you an arbitrary range but it does not have day resolution. Since we bill monthly it's generally not an issue

What about `sp_return_logs`, if we de-dupe by user?

Reply:
sp_return_logs can work but there is a lot of data in that one because of the dups so performance might be rough especially because we have to do multiple passes per sp.  i do envision that table as the future bottleneck and eventually we'll want to start archiving and/or limiting the queries to ranges.  i guess we can try it and see how it performs.

Reply:
Updated to use `sp_return_logs` and a raw SQL string in d7b24a30681f4dbe573aae6fb423de30c86472bb

This is already a "boil the ocean" query, and it's already running against the read replica, how important is optimizing it?

Reply:
update... I'm testing this out with prod data and it's timing out at 60s.... gonna try Steve's idea to group by dates

Reply:
We could use identities if we drop end date.  Since it runs daily they would just need to pull the report for the end date.